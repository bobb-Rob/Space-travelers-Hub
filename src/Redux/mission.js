import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const BASE_URL = 'https://api.spacexdata.com/v3/missions';
const MISSION_URL = BASE_URL;
const GET_MISSION = 'GET_MISSION';
const JOIN_MISSION = 'JOIN_MISSION';
const LEAVE_MISSION = 'LEAVE_MISSION';

function myData(data) {
  let newArr = [];
  data.forEach((mission) => {
    const obj = {};
    obj.id = mission.mission_id;
    obj.name = mission.mission_name;
    obj.description = mission.description;
    newArr = [...newArr, obj];
  });
  return newArr;
}

const initialState = [];

export const joiningMissionAction = (id) => ({
  type: JOIN_MISSION,
  payload: id,
});

export const leavingMissionAction = (id) => ({
  type: LEAVE_MISSION,
  payload: id,
});

export const getMission = createAsyncThunk(
  GET_MISSION,
  async () => {
    const response = await axios.get(MISSION_URL);
    return myData(response.data);
  },
);

const missionReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'GET_MISSION/fulfilled':
      return [...action.payload];
    case JOIN_MISSION: {
      const newState = state.map((mission) => {
        if (mission.id !== action.payload) { return mission; }
        return { ...mission, reserved: true };
      });
      return newState;
    }
    case LEAVE_MISSION: {
      const newState = state.map((mission) => {
        if (mission.id !== action.payload) { return mission; }
        return { ...mission, reserved: false };
      });
      return newState;
    }
    default:
      return state;
  }
};

export default missionReducer;
