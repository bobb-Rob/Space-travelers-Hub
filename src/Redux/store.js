import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import missionReducer from './mission';

const store = configureStore({
  reducer: {
    missionReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

export default store;
