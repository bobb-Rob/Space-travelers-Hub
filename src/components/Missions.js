import '../styles/missions.css';
import Mission from './Mission';

function Missions() {
  return (
    <div className="missions-list-container">
      <table className="table table-bordered table-sm">
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          <Mission />
        </tbody>
      </table>
    </div>
  );
}

export default Missions;
