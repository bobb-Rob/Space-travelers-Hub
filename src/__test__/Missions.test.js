import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Missions from '../components/Missions';
import store from '../Redux/store';
import { joiningMissionAction, leavingMissionAction } from '../Redux/mission';

describe('Missions tests', () => {
  it('render test', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should return an action with type JOIN_MISSION', () => {
    const id = 1;
    const action = joiningMissionAction(id);
    expect(action.type).toBe('JOIN_MISSION');
  });

  it('should should return an action with type LEAVE_MISSION', () => {
    const id = 1;
    const action = leavingMissionAction(id);
    expect(action.type).toBe('LEAVE_MISSION');
  });
});
