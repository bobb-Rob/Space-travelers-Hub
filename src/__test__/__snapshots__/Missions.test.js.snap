// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions tests render test 1`] = `
<div
  className="missions-list-container"
>
  <table
    className="table table-bordered table-sm"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
